### PR TITLE
Relativistic monte carlo

### PR DESCRIPTION
**Pull request by ben18785**
Addresses #953

### CODE REVIEW

File: pints/_mcmc/_relativistic.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+
+    def name(self):
+        """ See :meth:`pints.MCMCSampler.name()`. """
+        return 'Relativistic MCMC'
+
+    def needs_sensitivities(self):
+        """ See :meth:`pints.MCMCSampler.needs_sensitivities()`. """
+        return True
+
+    def _relativistic_mass(self):
```
Comment (MichaelClerx) [inline pints/_mcmc/_relativistic.py]: Single-use ~plastic~ function ?

File: pints/_mcmc/_relativistic.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+        """ Calculates relativistic mass. """
+        squared = np.sum(np.array(self._momentum)**2)
+        return self._mass * (squared / self._mc2 + 1)**0.5
```
Comment (MichaelClerx) [inline pints/_mcmc/_relativistic.py]: There's a chance the `sqrt()` method will be slightly faster / more accurate, but not a deal breaker!

File: pints/_mcmc/_relativistic.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+
+    def _log_init(self, logger):
+        """ See :meth:`Loggable._log_init()`. """
+        logger.add_float('Accept.')
+
+    def _log_write(self, logger):
+        """ See :meth:`Loggable._log_write()`. """
+        logger.log(self._mcmc_acceptance)
+
+    def _K(self, momentum):
```
Comment (MichaelClerx) [inline pints/_mcmc/_relativistic.py]: Maybe `_kinetic_energy` instead?

File: pints/_mcmc/_relativistic.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+        """
+        Kinetic energy or relativistic particle, which is defined in [1]_.
```
Comment (MichaelClerx) [inline pints/_mcmc/_relativistic.py]: `or` --> `of` ?

File: pints/_mcmc/_relativistic.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+
+    def set_mass(self, mass):
+        """ Sets scalar mass. """
+        if isinstance(mass, list):
+            raise ValueError('Mass must be scalar.')
+        if mass <= 0:
+            raise ValueError('Mass must be positive.')
+        self._mass = mass
+
+    def set_speed_light(self, c):
```
Comment (MichaelClerx) [inline pints/_mcmc/_relativistic.py]: Uhm... Maybe some quotes around "speed of light" in the docstring?

File: pints/_mcmc/_relativistic.py [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+        self._frog_iteration = 0
+
+        # Default number of leapfrog iterations
+        self._n_frog_iterations = 20
+
+        # Default integration step size for leapfrog algorithm
+        self._epsilon = 0.1
+        self._step_size = None
+        self._mass = 1
+        self._c = 10
```
Comment (MichaelClerx) [inline pints/_mcmc/_relativistic.py]: Don't let @chonlei see this

File: pints/_mcmc/_relativistic.py (hunk 5 above)
Comment (MichaelClerx) [inline pints/_mcmc/_relativistic.py]: `speed_of_light` (same below)

File: pints/_mcmc/_relativistic.py [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+            raise ValueError('Mass must be positive.')
+        self._mass = mass
+
+    def set_speed_light(self, c):
+        """ Sets speed of light. """
+        if c <= 0:
+            raise ValueError('Speed of light must be positive.')
+        self._c = c
+
+    def speed_light(self):
```
Comment (MichaelClerx) [inline pints/_mcmc/_relativistic.py]: `speed_of_light`

File: examples/sampling-relativistic-mcmc.ipynb [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
...spue2223j55ZczZDp48CAffPABr7zyCk888QQ7duzIKf8t\nt9zCmTNnsNvtKSGOU/N406YAACAASURBVMQvMaMk5gohHgG+DmR3sY2DaL5wLGd4RAhxCjNE/uRk\nx86GP6xxsMPF7KI5GOHJebBC6Cm9gy5cTGWsIpw12GXasEgdiZgSY2I4MJ46boUhXakfC79tds78\npeXGHE5ZChveNRWMIyUxBUF2MyS5dPtkyMidHIXIBIzxbB7BWM7VeEdLXvyRefaUihVrWWHModHi\noEHYxzgigUVkryw5EVrEMFePkYNXSPL1YD0IDAEngPuBdxjjx0tK2SalvDb6b4WU8pHo+w4p5Rel\nlEujj87RxlEoFIpPG83NzZw+naiQVl9fz6JFi+LGVE1NDV6vl9dee23cY3d2drJ//34AXn755Ywq\nf2vXrmXv3r20tpqeSr/fT0tLC16vF7fbzR133MHjjz9OfX19xtitra3ximh1dXWEw2HmzJlYGMl4\nkFK+CPxP4OeYBtF/kVL+rpBzCCEWA9cDH2fZtl4IcVgIcXhoaOIr5GVFpsLiLpmPPklFfqqNqyo5\n6QjMvCjUenO2ynnjIb1haTIzS7OvVRfJ6QtHGi+jFfTIhqtsIbk+jenzCeRHbcX03JuFIpglLwtg\nRtHUN8Ud718ZHSNuYBnjbNpbQTFCjs/TXWMdOx8xndHux4ZA9hYRU0G+VQQN4N+j/xSKC4qUgguu\n43RZMlcw7jSWTKNEiguN6S7I4fV6eeCBB3C5XNhsNpYsWcKmTZuorq7mu9/9LitXrmTx4sWsXj3+\nGg7Lly9ny5Yt3H///SxdupTvf//7Kdtra2vZvHkz3/rWtwiFzLCijRs3UllZyV133UUwGERKyWOP\nPZYx9uuvv87zzz9PUVERZWVl/Pa3v53OQhcDwB7M37UyIcQNUsq6QgwshKgAXgf+XkqZkdkerZC7\nCWDVqlUTtoyEEJQWWQiHBS4mn4M1ldRSnlGmfWoozPK1ZimjaBK5PUKmNkJN2ZbjHi/GmrPqYKEZ\n70p/Ph4sv2025Zq5Bj6aMj0eD8x0UOgwMCHMKLv5soJeUdgCHIaw5ayKWGSzwATzCfMlV9GWiKUs\nq8fSKyLUC9ObnOv7EB/bUoLNSOS8WRCUYCU4jpYGC0UVgwX8Wzhez+1kyLeK
```
Comment (MichaelClerx) [inline examples/sampling-relativistic-mcmc.ipynb]: Monomial!

File: pints/_mcmc/_relativistic.py [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+            self._position = np.array(self._current, copy=True)
+            self._gradient = np.array(self._current_gradient, copy=True)
+            self._momentum = np.array(self._current_momentum, copy=True)
+
+            # Perform a half-step before starting iteration 0 below
+            self._momentum -= self._scaled_epsilon * self._gradient * 0.5
+
+        # Perform a leapfrog step for the position
+        self._position += (
+            self._scaled_epsilon * (self._relativistic_mass()**(-1)) *
```
[the remark this answers is not in the document]
Comment (ben18785) [reply pints/_mcmc/_relativistic.py]: Done!

File: pints/_mcmc/_relativistic.py [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+
+    def _K(self, momentum):
+        """
+        Kinetic energy or relativistic particle, which is defined in [1]_.
+        """
+        squared = np.sum(np.array(momentum)**2)
+        return self._mc2 * (squared / self._mc2 + 1)**0.5
+
+    def mass(self):
+        """ Returns mass. """
```
[the remark this answers is not in the document]
Comment (ben18785) [reply pints/_mcmc/_relativistic.py]: Done

File: pints/_mcmc/_relativistic.py (hunk 1 above)
Comment (ben18785) [reply pints/_mcmc/_relativistic.py]: Yep, removed!

File: pints/_mcmc/_relativistic.py (hunk 2 above)
Comment (ben18785) [reply pints/_mcmc/_relativistic.py]: Yep, changed.

File: pints/_mcmc/_relativistic.py (hunk 3 above)
Comment (ben18785) [reply pints/_mcmc/_relativistic.py]: Done!

File: pints/_mcmc/_relativistic.py (hunk 5 above)
Comment (ben18785) [reply pints/_mcmc/_relativistic.py]: Haha, yep. I don't want to be playing God. Done.

File: pints/_mcmc/_relativistic.py (hunk 6 above)
Comment (ben18785) [reply pints/_mcmc/_relativistic.py]: Haha, I'm a physicist too and quite enjoy it.

File: pints/_mcmc/_relativistic.py (hunk 5 above)
Comment (ben18785) [reply pints/_mcmc/_relativistic.py]: Done.

File: pints/_mcmc/_relativistic.py (hunk 7 above)
Comment (ben18785) [reply pints/_mcmc/_relativistic.py]: Done.

File: examples/sampling-relativistic-mcmc.ipynb (hunk 8 above)
Comment (ben18785) [reply examples/sampling-relativistic-mcmc.ipynb]: Good spot!

File: pints/_mcmc/_relativistic.py (hunk 4 above)
Comment (MichaelClerx) [reply pints/_mcmc/_relativistic.py]: Kinetic energy _of_ particle?

**Comment by MichaelClerx** [comment]
One tiny one left! (Shows up "hidden" by default for me)

File: pints/_mcmc/_relativistic.py (hunk 4 above)
Comment (ben18785) [reply pints/_mcmc/_relativistic.py]: Done!

**Comment by ben18785** [comment]
@MichaelClerx Done! Feel free to merge if you're happy with it.

**Pull request merged**